react/components/EditorContainer/Sidebar/ComponentList/utils.ts

```diff
@@ -15,97 +17,97 @@ export const getParentTreePath = (treePath: string): string => {
   return splitTreePath.slice(0, splitTreePath.length - 1).join('/')
 }
 
-export const isRootComponent = (component: SidebarComponent) => {
+export const isRootComponent = (minimumTreePathSize: number) => (
+  component: SidebarComponent
+) => {
   const splitTreePath = component.treePath.split('/')
 
   return (
-    splitTreePath.length === 2 ||
-    (splitTreePath.length === 3 && splitTreePath[2].startsWith('$'))
+    splitTreePath.length === minimumTreePathSize ||
+    (splitTreePath.length === minimumTreePathSize + 1 &&
+      splitTreePath[minimumTreePathSize].startsWith('$'))
   )
 }
 
-export const isRootComponentRelative = (rootComponent: SidebarComponent) => {
+export const isRootComponentRelative = (minimumTreePathSize: number) => (
+  rootComponent: SidebarComponent
+) => {
   const splitTreePath = rootComponent.treePath.split('/')
 
   return (
-    splitTreePath[1].startsWith('$') ||
-    (splitTreePath.length === 3 && splitTreePath[2].startsWith('$'))
+    splitTreePath[minimumTreePathSize - 1].startsWith('$') ||
+    (splitTreePath.length === minimumTreePathSize + 1 &&
+      splitTreePath[minimumTreePathSize].startsWith('$'))
   )
 }
 
-export const normalize = (components: SidebarComponent[]) => {
-  const [roots, leaves] = partition(isRootComponent, components)
-
-  const [relativeRoots, nonRelativeRoots] = partition(
-    isRootComponentRelative,
-    roots
+const mountFullTree = (
+  root: SidebarComponent,
+  rootOwnLeaves: SidebarComponent[]
+): NormalizedComponent => {
+  if (rootOwnLeaves.length === 0) {
+    return { ...root, isSortable: false }
+  }
+  const rootTreePathLength = pathLength(root.treePath)
+  const immediateChildren = rootOwnLeaves.filter(
+    ({ treePath: childPath }) =>
+      rootTreePathLength + 1 === pathLength(childPath)
+  )
+  const nonImmediateChildren = rootOwnLeaves.filter(
+    ({ treePath: childPath }) => rootTreePathLength + 1 < pathLength(childPath)
   )
 
-  const normalizedRelativeRoots = relativeRoots.reduce<NormalizedRelativeRoot>(
-    (acc, currRoot) => {
-      const splitTreePath = currRoot.treePath.split('/')
-
-      const treePathTail = splitTreePath[splitTreePath.length - 1]
-
-      const blockRole = getBlockRole(treePathTail)
+  return {
+    ...root,
+    components: immediateChildren.map(child => {
+      const grandChildren = nonImmediateChildren.filter(({ treePath }) =>
```

Comment:
:pick: Grandchildren can't have more than one parent, so the `filter` iterates through elements that will evaluate to `true` N - 1 times. I'm just pointing it out though, since the `normalize` function only runs once during the list's entire life cycle and the number of components is usually relatively small.